src/SignalR/clients/csharp/Client.SourceGenerator/src/HubProxyGenerator.Emitter.cs

```diff
@@ -0,0 +1,169 @@
+            private readonly SourceGenerationSpec _spec;
+
+            public Emitter(GeneratorExecutionContext context, SourceGenerationSpec spec)
+            {
+                _context = context;
+                _spec = spec;
+            }
+
+            public void Emit()
+            {
+                EmitExtensions();
+                foreach (var classSpec in _spec.Classes)
+                {
+                    EmitProxy(classSpec);
+                }
+            }
+
+            private void EmitExtensions()
+            {
+                var getProxyBody = new StringBuilder();
+
+                foreach (var classSpec in _spec.Classes)
+                {
+                    var fqIntfTypeName = classSpec.FullyQualifiedInterfaceTypeName;
+                    var fqClassTypeName = classSpec.FullyQualifiedClassTypeName;
+                    getProxyBody.Append($@"
+            if(typeof(THub) == typeof({fqIntfTypeName})) return (THub) ({fqIntfTypeName}) new {fqClassTypeName}(conn);");
+                }
+
+                var getProxy = $@"// Generated by Microsoft.AspNetCore.Client.SourceGenerator
+using Microsoft.AspNetCore.SignalR.Client;
+namespace Microsoft.AspNetCore.SignalR.Client
+{{
+    public static partial class HubConnectionExtensionsGenerated
+    {{
+        public static THub GetProxy<THub>(this IHubConnection conn)
+        {{
+{getProxyBody.ToString()}
+            throw new System.ArgumentException(nameof(THub));
+        }}
+    }}
+}}";
+
+                _context.AddSource("HubProxy.g.cs", SourceText.From(getProxy.ToString(), Encoding.UTF8));
+            }
+
+            private void EmitProxy(ClassSpec classSpec)
+            {
+                var methods = new StringBuilder();
+
+                foreach (var methodSpec in classSpec.Methods)
+                {
+                    var signature = new StringBuilder($"public {methodSpec.FullyQualifiedReturnTypeName} {methodSpec.Name}(");
+                    var callArgs = new StringBuilder("");
+                    var signatureArgs = new StringBuilder("");
+                    var first = true;
+                    foreach (var argumentSpec in methodSpec.Arguments)
+                    {
+                        if (!first)
+                        {
+                            signatureArgs.Append(", ");
+                        }
+
+                        first = false;
+                        signatureArgs.Append($"{argumentSpec.FullyQualifiedTypeName} {argumentSpec.Name}");
+                        callArgs.Append($", {argumentSpec.Name}");
+                    }
+                    signature.Append(signatureArgs);
+                    signature.Append(')');
+
+                    // Prepare method body
+                    var body = "";
+                    if (methodSpec.Support != SupportClassification.Supported)
+                    {
+                        body = "throw new System.NotSupportedException();";
```

Comment:
Not a bad idea, warn during source generation, but still allow the generation to run so users can still use other methods.

Can we flow a specific error message through and add that to the exception?

Reply:
Definitely. I intended method spec to be sufficient to re-construct whatever needed but we can add a string field with a specific message.